### ISSUE TITLE
Directional formatting disrupts accessible strings

### ISSUE THREAD

**Issue by zepumph** [phetsims/chipper#798]
Since moving a11y strings to support nested objects (still in a branch as of this writing see https://github.com/phetsims/chipper/issues/777), as a part of https://github.com/phetsims/rosetta/issues/193, we are now surrounding all a11y strings in directional formatting. While working on https://github.com/phetsims/friction/issues/182, I found that this is detrimental to the UX with a screen reader. This is because it adds these all over, including many times in the middle of strings (since we fill in so many values with sub-strings/clauses). 

Right now I decided to have a11y strings opt-out of directional formatting, but we will need to figure out a way around this before a11y strings will support i18n. I don't think this is currently high priority.

One thing we could do is somehow make it so that the directional formatting is applied more downstream, like when a string is "sent to the sim" instead of loaded from a JSON. Maybe this would look like wrapping usages in Text.set, and some setters in Accessibility.js (PDOM.js). That sounds gross. I'm not looking forward to this issue.

**Comment by zepumph** [comment]
Tagging @jessegreenberg just so he is aware about this.

**Comment by zepumph** [comment]
This needs to block any adding of a11y strings into strings files, so I'm going to assign people now that the support is in chipper.

**Comment by zepumph** [comment]
Over in https://github.com/phetsims/scenery-phet/issues/570 we ran into a case where using StringUtils to add one string into another made caused a spacing error in rich text. The case is as follows, take this string:

`"Move to next item{{orGroup}}" `

which can get filled in to be:


`"Move to next item or group" `

such that there is another translatable string getting filled in: " or group"

The directional formatting on the beginning of " or group" caused that next space to not show up in the RichText the string was provided to; it looked like:


`"Move to next itemor group" `

I have marked a TODO in `GeneralKeyboardHelpSection` regarding this because to solve it, I am manually removing the directional formatting. Marking high priority.

I think that a dev meeting is in order to sort this out. I'm not sure how to navigate around this. Perhaps we need to reevaluate how we are combining translatable strings.

**Comment by jonathanolson** [comment]
Can we just remove directional marks when the a11y strings are delivered to users through the DOM/etc.?

**Comment by jessegreenberg** [comment]
@jonathanolson showed us StringUtils.stripEmbeddingMarks which may help remove embedding marks before we provide text for a screen reader.

**Comment by zepumph** [comment]
>@jonathanolson showed us `StringUtils.stripEmbeddingMarks` which may help remove embedding marks before we provide text for a screen reader.

I think that we can apply this to all interactive descriptions as a hold over pattern until we need to worry about i18n with a11y strings.

In general we do not want to strip directional formatting, like done in https://github.com/phetsims/chipper/issues/798#issuecomment-572329507. Instead @jonathanolson said that it would be best to handle each buggy case with directional formatting, if possible, and then if not strip the unicode characters very specifically and purposefully. 

For the `RichText` bug, I will create another issue and work with @jonathanolson to try to fix the spacing issue.

**Comment by zepumph** [comment]
For `stripEmbeddingMarks `, @jessegreenberg how do you think we should target this? We could for now put it in two places (one in utterance queue and one in scenery) with TODOs pointing to this issue. 

I didn't make a patch, but I did a bit of investigation. This would involve adding phetcommon as a dependency to utterance queue (which feels fine to me) and to scenery (not sure about this one).

How does that sound?

**Comment by jessegreenberg** [comment]
I am a little hesitant to add a dependency to scenery, @jonathanolson how do you feel about adding phetcommon to scenery? I am wondering if this is a PhET specific decision and so shouldn't be in scenery, does this belong in joist?

Just brainstorming alternatives - 
 - Is `stripEmbeddingMarks` fast enough to be called frequently on the entire PDOM, like every step or something? Likely not.
- Perhaps we can use [MutationObserver](https://davidwalsh.name/mutationobserver-api) to strip embedding marks from a specific added/changed node/attributes in the PDOM.
- Or put a copy of `stripEmbeddingMarks` in AccessibilityUtils so we don't have to bring in phetcommon?

These could be possibilities if it is not OK to pull phetcommon into scenery.

**Comment by jessegreenberg** [comment]
I asked @jonathanolson over slack and he said making phetcommon a scenery dependency was workable, or alternatively we could move `stripEmbeddingMarks` to phet-core.

@zepumph of options so far do you have a preference? Thoughts on if this should be in scenery or joist? I don't feel strongly.

**Comment by jessegreenberg** [comment]
`stripEmbeddingMarks` is only used in one spot in PhetMenu right now, my preference is to move the function to phet-core.

**Comment by zepumph** [comment]
This is great! Moved to phet-core in the above commit. @jessegreenberg do you want to review?

Should we keep this issue open and unassigned as a child issue for work when we do https://github.com/phetsims/rosetta/issues/221? (I think we should)

Do you want to cherry-pick this workaround into the Molarity rc branch?

**Comment by zepumph** [comment]
Now that we are covering all the end points, I removed the workaround at the source for a11y strings.

**Comment by zepumph** [comment]
Actually I could get rid of the utterance queue usage of `stripEmbeddingMarks` because AriaHerald uses AccessibilityUtils.setTextContent.

**Comment by jessegreenberg** [comment]
These changes are great! I also verified that embedding marks that used to be in a PDOM around translatable strings are no longer there.

> Do you want to cherry-pick this workaround into the Molarity rc branch?

Thanks, I will, there are some marks around PDOM strings and it would be beneficial to remove them. Assigning to myself for this.

> Should we keep this issue open and unassigned as a child issue for work when we do phetsims/rosetta#221? (I think we should)

Yes, sounds good.

**Comment by zepumph** [comment]
Once review is complete we can demote priority and mark https://github.com/phetsims/chipper/issues/795 off hold.

**Comment by jessegreenberg** [comment]
This has been cherry-picked into molarity rc branch.

This is on hold now until i18n for a11y strings is supported.

**Comment by jessegreenberg** [comment]
AriaHerald currently uses PDOMUtils.setTextContent which uses stripEmbeddingMarks so it seems OK for now. Back on hold, feel free to remove assignment if you agree.

**Comment by jessegreenberg** [comment]
This is coming up again for aria-valuetext (and probably other PDOM attributes) in https://github.com/phetsims/chipper/issues/798. I think we need to use `stripEmbeddingMarks` when setting attributes somehow.

**Comment by zepumph** [comment]
Good idea! Presumably there would only be a few places in PDOMPeer where we can do this. Like `setAttributeToElement`.

**Comment by jessegreenberg** [comment]
Thanks! I just reviewed functions in PDOMPeer and `setAttributeToElement` seems like the only one that needs this. Other places where DOM API is used directly is setting things that would never have directional formatting or ever be translatable as far as I can tell. I added `stripEmbeddingMarks` to `setAttributeToElement` in the above commit, feel free to take a look if you wish otherwise this is back on hold.

**Comment by phet-dev** [comment]
Reopening because there is a TODO marked for this issue.
